File: app/src/main/java/com/digis/android/task/data/source/remote/RemoteCallback.java
package com.digis.android.task.data.source.remote;

import androidx.annotation.NonNull;

/**
 * Communicates responses from a remote request.
 *
 * @param <T> Successful response type.
 */
public interface RemoteCallback<T> {
    /**
     * Invoked when the request has been succeeded.
     *
     * @param httpResponse The response
     */
    void onResponse(@NonNull HttpResponse<T> httpResponse);

    /**
     * Invoked when the request has been failed.
     *
     * @param throwable The throwable
     */
    void onFailure(Throwable throwable);

    /**
     * Invoked when the request has been canceled.
     */
    default void onCanceled() { }
}

File: app/src/main/java/com/digis/android/task/implementation/kotlin/data/source/remote/RemoteCallback.kt
package com.digis.android.task.implementation.kotlin.data.source.remote

/**
 * Communicates responses from a remote request.
 *
 * @param <T> Successful response type.
 */
interface RemoteCallback<T> {
    /**
     * Invoked with [httpResponse] when the request has been succeeded.
     */
    fun onResponse(httpResponse: HttpResponse<T>)

    /**
     * Invoked with [throwable] when the request has been failed.
     */
    fun onFailure(throwable: Throwable)

    /**
     * Invoked when the request has been canceled.
     */
    fun onCanceled() { }
}